pkg/destroy/nutanix/nutanix.go

```diff
@@ -0,0 +1,229 @@
+	nutanixclientv3 "github.com/terraform-providers/terraform-provider-nutanix/client/v3"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/installer/pkg/destroy/providers"
+	installertypes "github.com/openshift/installer/pkg/types"
+	nutanixtypes "github.com/openshift/installer/pkg/types/nutanix"
+)
+
+const (
+	emptyFilter                 = ""
+	expectedCategoryKeyFormat   = "kubernetes-io-cluster-%s"
+	expectedCategoryValueOwned  = "owned"
+	expectedCategoryValueShared = "shared"
+)
+
+// clusterUninstaller holds the various options for the cluster we want to delete.
+type clusterUninstaller struct {
+	clusterID string
+	infraID   string
+	v3Client  *nutanixclientv3.Client
+	logger    logrus.FieldLogger
+}
+
+type expectedCategory struct {
+	key    string
+	values []string
+}
+
+// New returns an Nutanix destroyer from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *installertypes.ClusterMetadata) (providers.Destroyer, error) {
+	v3Client, err := nutanixtypes.CreateNutanixClient(context.TODO(),
+		metadata.ClusterPlatformMetadata.Nutanix.PrismCentral,
+		metadata.ClusterPlatformMetadata.Nutanix.Port,
+		metadata.ClusterPlatformMetadata.Nutanix.Username,
+		metadata.ClusterPlatformMetadata.Nutanix.Password,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &clusterUninstaller{
+		clusterID: metadata.ClusterID,
+		infraID:   metadata.InfraID,
+		v3Client:  v3Client,
+		logger:    logger,
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *clusterUninstaller) Run() (*installertypes.ClusterQuota, error) {
+	o.logger.Infof("Starting deletion of Nutanix infrastructure for Openshift cluster %s", o.infraID)
+	err := wait.PollImmediateInfinite(time.Second*30, o.destroyCluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to destroy cluster")
+	}
+
+	return nil, nil
+}
+
+func (o *clusterUninstaller) destroyCluster() (bool, error) {
+	cleanupFuncs := []struct {
+		name    string
+		execute func(*clusterUninstaller) error
+	}{
+		{name: "VMs", execute: cleanupVMs},
+		{name: "Images", execute: cleanupImages},
+		{name: "Categories", execute: cleanupCategories},
+	}
+
+	done := true
+	for _, cleanupFunc := range cleanupFuncs {
+		if err := cleanupFunc.execute(o); err != nil {
+			o.logger.Debugf("%s: %q", cleanupFunc.name, err)
+			done = false
+		}
+	}
+	return done, nil
+}
+
+func cleanupVMs(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	matchedVirtualMachineList := make([]*nutanixclientv3.VMIntentResource, 0)
+	allVMs, err := o.v3Client.V3.ListAllVM(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range allVMs.Entities {
+		if hasCategoryAssigned(v.Metadata, ec) {
+			matchedVirtualMachineList = append(matchedVirtualMachineList, v)
+		}
+	}
+
+	if len(matchedVirtualMachineList) == 0 {
+		o.logger.Infof("no VMs found that require deletion for cluster %s", o.clusterID)
```

Comment:
Log messages should start with a capital letter.
```suggestion
		o.logger.Infof("No VMs found that require deletion for cluster %s", o.clusterID)
```